platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateFileWriter.java

```diff
@@ -135,27 +143,142 @@ public static void writeStateFile(final Path directory, final SignedState signed
+
+        final boolean copyPreconsensusStream = platformContext
+                .getConfiguration()
+                .getConfigData(PreconsensusEventStreamConfig.class)
+                .copyRecentStreamToStateSnapshots();
+        if (!copyPreconsensusStream) {
+            // PCES copying is disabled
+            return;
+        }
+
+        // The PCES files will be copied into this directory
+        final Path pcesDestination =
+                destinationDirectory.resolve("preconsensus-events").resolve(Long.toString(selfId.id()));
+        Files.createDirectories(pcesDestination);
+
+        // Gather all PCES files currently on disk.
+        final List<PreconsensusEventFile> allFiles = new ArrayList<>();
+        final Path preconsensusEventStreamDirectory =
+                PreconsensusEventFileManager.getDatabaseDirectory(platformContext, selfId);
+        try (final Stream<Path> stream = Files.walk(preconsensusEventStreamDirectory)) {
+            stream.filter(Files::isRegularFile).forEach(path -> {
+                try {
+                    allFiles.add(PreconsensusEventFile.of(path));
+                } catch (final IOException e) {
+                    // Ignore, this will get thrown for each file that is not a PCES file
+                }
+            });
+        }
+
+        if (allFiles.isEmpty()) {
+            logger.warn(STATE_TO_DISK.getMarker(), "No preconsensus event files found to copy");
+            return;
+        }
+
+        // Sort by sequence number
+        Collections.sort(allFiles);
+
+        // Discard all files that either have an incorrect origin or that do not contain non-ancient events
+        final List<PreconsensusEventFile> filesToCopy = new ArrayList<>();
+        final PreconsensusEventFile lastFile = allFiles.get(allFiles.size() - 1);
+        for (final PreconsensusEventFile file : allFiles) {
+            if (file.getOrigin() == lastFile.getOrigin()
+                    && file.getMaximumGeneration() <= minimumGenerationNonAncient) {
+                filesToCopy.add(file);
+            }
+        }
+
+        if (filesToCopy.isEmpty()) {
+            logger.warn(
+                    STATE_TO_DISK.getMarker(), "No preconsensus event files meeting specified criteria found to copy");
+            return;
+        }
+
+        logger.info(
+                STATE_TO_DISK.getMarker(),
+                "Copying {} preconsensus event files to state snapshot directory",
+                filesToCopy.size());
+
+        // Although the last file may be currently in the process of being written, all previous files will
+        // be closed and immutable and so it's safe to hard link them.
+        for (int index = 0; index < filesToCopy.size() - 1; index++) {
+            final PreconsensusEventFile file = filesToCopy.get(index);
+            final Path destination = pcesDestination.resolve(file.getFileName());
+            try {
+                Files.createLink(destination, file.getPath());
```

Comment:
so if the original files are deleted, these links are broken?

Reply:
This creates a hard link. Think of a hard link like a reference to an object in java. You aren't copying the file, you just copy a reference to a file. As long as there is at least one reference to the file, then the OS won't actually delete the file contents.

If this were a symbolic link, it would be a different story. Symbolic links are analogous to weak references in java. It doesn't copy the data, but it also doesn't prevent GC on the underlying object/file.